### PR TITLE
SONARPY-154 False Positive on rules considering nested class properties

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/NewSymbolsAnalyzer.java [hunk 1]
```diff
@@ -36,7 +37,10 @@
     List<AstNode> methods = classDef.getFirstChild(PythonGrammar.SUITE).getDescendants(PythonGrammar.FUNCDEF);
 
     for (AstNode method : methods) {
-      addFieldsInMethod(method);
+      // ignore any nested class
+      if (classDef == method.getFirstAncestor(PythonGrammar.CLASSDEF)) {
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/NewSymbolsAnalyzer.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: python-checks/src/main/java/org/sonar/python/checks/NewSymbolsAnalyzer.java (hunk 1 above)
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/NewSymbolsAnalyzer.java]: OK Mr SonarTech (with bad will)

File: python-checks/src/test/java/org/sonar/python/checks/utils/AstNodeUtils.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+public class AstNodeUtils {
```
Comment (vilchik-elena) [inline python-checks/src/test/java/org/sonar/python/checks/utils/AstNodeUtils.java]: @ivandalbosco We should remove the code which is not used at the moment

Comment (ivandalbosco) [inline python-checks/src/test/java/org/sonar/python/checks/utils/AstNodeUtils.java]: @vilchik-elena Generally speaking, I agree. But not this one, please!

Comment (ivandalbosco) [inline python-checks/src/test/java/org/sonar/python/checks/utils/AstNodeUtils.java]: @vilchik-elena Done. God bless PythonToolkit!

File: python-checks/src/test/resources/checks/fieldDuplicatesClassName.py [hunk 3]
```diff
@@ -14,6 +14,12 @@ class NestedClass:
     def fun(self):
       self.nestedClass = 5 # Noncompliant [[secondary=-2]] {{Rename field "nestedClass"}}
 #          ^^^^^^^^^^^
+      myClass = 1
```
Comment (vilchik-elena) [inline python-checks/src/test/resources/checks/fieldDuplicatesClassName.py]: @ivandalbosco is it intended to test local variable?

File: python-checks/src/test/resources/checks/fieldDuplicatesClassName.py [hunk 4]
```diff
@@ -14,6 +14,12 @@ class NestedClass:
...
 
 class MyClass1(MyClass):
   myClass1 = 1
+
+class MyClass:
+  class B:
+     def foo(self):
+         self.myClass = 1
```
Comment (vilchik-elena) [inline python-checks/src/test/resources/checks/fieldDuplicatesClassName.py]: @ivandalbosco no unit test for the field declared in class body
```
class MyClass:
  class B:
     myClass = 1 # FP
```

File: python-checks/src/test/resources/checks/fieldDuplicatesClassName.py (hunk 3 above)
Comment (ivandalbosco) [inline python-checks/src/test/resources/checks/fieldDuplicatesClassName.py]: @vilchik-elena yes. It checks a false positive on the enclosing class MyClass

Comment (vilchik-elena) [inline python-checks/src/test/resources/checks/fieldDuplicatesClassName.py]: @ivandalbosco but there is no FP here as it's local variable, not field. You should prepend this code with `self.` to have FP (and btw this case is already tested on line 25).

Comment (ivandalbosco) [inline python-checks/src/test/resources/checks/fieldDuplicatesClassName.py]: @vilchik-elena yep, you are right

**Pull request merged**